File: learnwithme/project999/Til.hs
-- Facebook friend who post the following & <Mr. Robot Season 2, Episode 11>, ROT13 cipher   
-- trigger me to create a Haskell script to comb list of english words that fulfil the requirement
-- https://www.reddit.com/r/todayilearned/comments/2x3gea/til_that_if_you_take_all_the_letters_from_the/

import Data.Char
import Network.HTTP.Conduit
import Network.HTTP
import qualified Data.ByteString.Lazy as LB (unpack)
import Data.Maybe (catMaybes)

revrot :: Char -> Char
revrot x     
    | charint >= 97 && charint <= 110 = chr $ 122 - (charint - 97)
    | charint > 111 && charint <= 122 = chr $ 97 + (122 - charint)
    | otherwise = x
    where charint = Data.Char.ord $ toLower x

revword :: String -> String
revword x = reverse $ map revrot x

checkrevword :: String -> Maybe (String, String, String)
checkrevword x
    | x == y = Just (x,z,y)
    | otherwise = Nothing
    where y = revword x
          z = map revrot x

-- Ref https://www.quora.com/Where-can-I-find-a-dictionary-in-a-CSV-or-Excel-format
url :: String
url = "https://raw.githubusercontent.com/eneko/data-repository/master/data/words.txt"

dlDictionaryCsv ::  IO [String]
dlDictionaryCsv = do 
                  return.lines.bsToStr =<< simpleHttp url
    where bsToStr = map (chr . fromEnum).LB.unpack

main :: IO ()
main = do    
    -- lWords <- (return.lines) =<< readFile "words.csv" -- Temp for local reading
    lWords <- dlDictionaryCsv
    let lValidReqWords = catMaybes $ map checkrevword lWords
    mapM_ print lValidReqWords
    

File: learnwithme/project999/Til.py
# Facebook friend who post the following & <Mr. Robot Season 2, Episode 11>, ROT13 cipher   
# trigger me to create a Haskell script to comb list of english words that fulfil the requirement
# https://www.reddit.com/r/todayilearned/comments/2x3gea/til_that_if_you_take_all_the_letters_from_the/
# Original Haskell version: https://github.com/adamliu84/KTSWorkshop/blob/master/learnwithme/project999/Til.hs

import urllib.request
import csv
import io

def parseOnlineCSV():
    url = "https://raw.githubusercontent.com/eneko/data-repository/master/data/words.txt"
    pageResponse = urllib.request.urlopen(url)
    awords = csv.reader(io.TextIOWrapper(pageResponse))
    return awords

def checkrevword(temp):
    word = temp[0].lower()
    revword = ''.join(list(map(revrot, word)))
    compareword = revword[::-1] #http://stackoverflow.com/questions/931092/reverse-a-string-in-python  
    return (word == compareword)

def revrot(singlechar):
    charord = ord(singlechar)
    if charord >= 97 and charord <= 110:
        return(chr(122 - (charord-97)))
    elif charord > 111 and charord <= 122:
        return(chr(97 + (122-charord)))
    else:
        return(singlechar)
    
awords = parseOnlineCSV()
aresult = filter(checkrevword, awords)
for row in aresult:
    print(row)